// @engie-group/fluid-design-system/lib-esm/components/form/search-input.js
/**
 * --------------------------------------------------------------------------
 * NJ: search-input.ts
 * --------------------------------------------------------------------------
 */
import AbstractComponent from '../../globals/ts/abstract-component';
import { Core } from '../../globals/ts/enum';
import Data from '../../globals/ts/data';
import EventHandler from '../../globals/ts/event-handler';
import Manipulator from '../../globals/ts/manipulator';
import WebComponentFactory from '../../globals/ts/web-component-factory';
export default class SearchInput extends AbstractComponent {
    constructor(element, options = {}) {
        super(SearchInput, element, Manipulator.extend(true, SearchInput.DEFAULT_OPTIONS, options));
        this.setListeners();
        Data.setData(element, SearchInput.DATA_KEY, this);
    }
    static init(options = {}) {
        return super.init(this, options, SearchInput.SELECTOR.default);
    }
    dispose() {
        Data.removeData(this.element, SearchInput.DATA_KEY);
        this.element = null;
    }
    static getInstance(element) {
        return Data.getData(element, SearchInput.DATA_KEY);
    }
    setListeners() {
        const input = this.element.getElementsByClassName(SearchInput.INPUT_CLASS)[0];
        const resetButton = this.element.getElementsByClassName(SearchInput.RESET_CLASS)[0];
        EventHandler.on(resetButton, 'click', () => {
            input.value = null;
            this.element.classList.remove('is-filled');
        });
    }
}
SearchInput.NAME = `${Core.KEY_PREFIX}-form-input-search`;
SearchInput.DATA_KEY = `${Core.KEY_PREFIX}.search-input`;
SearchInput.SELECTOR = {
    default: `.${SearchInput.NAME}`
};
SearchInput.INPUT_CLASS = `${Core.KEY_PREFIX}-form-control`;
SearchInput.RESET_CLASS = `${Core.KEY_PREFIX}-form-control__reset`;
SearchInput.DEFAULT_OPTIONS = {
    selector: SearchInput.SELECTOR
};
export class SearchInputWC extends WebComponentFactory {
    constructor() {
        super(SearchInput);
    }
    static init() {
        WebComponentFactory.init(SearchInputWC);
    }
}
SearchInputWC.TAG_NAME = SearchInput.NAME;
