// @engie-group/fluid-design-system/lib-esm/globals/ts/enum/native-event.js
export var NativeEvent;
(function (NativeEvent) {
    NativeEvent["click"] = "click";
    NativeEvent["dblclick"] = "dblclick";
    NativeEvent["mouseup"] = "mouseup";
    NativeEvent["mousedown"] = "mousedown";
    NativeEvent["contextmenu"] = "contextmenu";
    NativeEvent["mousewheel"] = "mousewheel";
    NativeEvent["DOMMouseScroll"] = "DOMMouseScroll";
    NativeEvent["mouseover"] = "mouseover";
    NativeEvent["mouseout"] = "mouseout";
    NativeEvent["mousemove"] = "mousemove";
    NativeEvent["selectstart"] = "selectstart";
    NativeEvent["selectend"] = "selectend";
    NativeEvent["keydown"] = "keydown";
    NativeEvent["keypress"] = "keypress";
    NativeEvent["keyup"] = "keyup";
    NativeEvent["orientationchange"] = "orientationchange";
    NativeEvent["touchstart"] = "touchstart";
    NativeEvent["touchmove"] = "touchmove";
    NativeEvent["touchend"] = "touchend";
    NativeEvent["touchcancel"] = "touchcancel";
    NativeEvent["pointerdown"] = "pointerdown";
    NativeEvent["pointermove"] = "pointermove";
    NativeEvent["pointerup"] = "pointerup";
    NativeEvent["pointerleave"] = "pointerleave";
    NativeEvent["pointercancel"] = "pointercancel";
    NativeEvent["gesturestart"] = "gesturestart";
    NativeEvent["gesturechange"] = "gesturechange";
    NativeEvent["gestureend"] = "gestureend";
    NativeEvent["focus"] = "focus";
    NativeEvent["blur"] = "blur";
    NativeEvent["change"] = "change";
    NativeEvent["reset"] = "reset";
    NativeEvent["select"] = "select";
    NativeEvent["submit"] = "submit";
    NativeEvent["focusin"] = "focusin";
    NativeEvent["focusout"] = "focusout";
    NativeEvent["load"] = "load";
    NativeEvent["unload"] = "unload";
    NativeEvent["beforeunload"] = "beforeunload";
    NativeEvent["resize"] = "resize";
    NativeEvent["move"] = "move";
    NativeEvent["DOMContentLoaded"] = "DOMContentLoaded";
    NativeEvent["readystatechange"] = "readystatechange";
    NativeEvent["error"] = "error";
    NativeEvent["abort"] = "abort";
    NativeEvent["scroll"] = "scroll";
})(NativeEvent || (NativeEvent = {}));
