// @engie-group/fluid-design-system/lib-esm/components/navbar/index.js
/**
 * --------------------------------------------------------------------------
 * NJ : Navbar.ts
 * --------------------------------------------------------------------------
 */
import AbstractComponent from '../../globals/ts/abstract-component';
import CollapseComponent from '../collapse';
import CollapseSearchBar from './collapse-searchbar';
import { Core } from '../../globals/ts/enum';
import Data from '../../globals/ts/data';
import EventHandler from '../../globals/ts/event-handler';
import WebComponentFactory from '../../globals/ts/web-component-factory';
export default class Navbar extends AbstractComponent {
    constructor(element) {
        super(Navbar, element);
        this.registerEvents();
    }
    dispose() {
        Data.removeData(this.element, Navbar.DATA_KEY);
        this.element = null;
    }
    handleCollapseShow() {
        this.element.classList.add(Navbar.CLASS_NAME.shownCollapse);
    }
    handleCollapseHide() {
        this.element.classList.remove(Navbar.CLASS_NAME.shownCollapse);
    }
    static getInstance(element) {
        return Data.getData(element, Navbar.DATA_KEY);
    }
    static init(options = {}) {
        CollapseSearchBar.init(); // Init CollapseSearchBar if its selector exists
        return super.init(this, options, Navbar.SELECTOR.default);
    }
    /**
     * ------------------------------------------------------------------------
     * Data Api implementation
     * ------------------------------------------------------------------------
     */
    registerEvents() {
        EventHandler.on(this.element, CollapseComponent.EVENT.show, () => {
            this.handleCollapseShow();
        });
        EventHandler.on(this.element, CollapseComponent.EVENT.hidden, () => {
            this.handleCollapseHide();
        });
    }
}
Navbar.NAME = `${Core.KEY_PREFIX}-navbar`;
Navbar.DATA_KEY = `${Core.KEY_PREFIX}.navbar`;
Navbar.CLASS_NAME = {
    shownCollapse: `${Navbar.NAME}--shown-collapse`
};
Navbar.SELECTOR = {
    default: `.${Navbar.NAME}`
};
export class NavbarWC extends WebComponentFactory {
    constructor() {
        super(Navbar, CollapseComponent);
    }
    static init() {
        WebComponentFactory.init(NavbarWC);
    }
}
NavbarWC.TAG_NAME = Navbar.NAME;
